File: Medium/173. BST Iterator/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class BSTIterator {

    private Queue<TreeNode> q;
    public BSTIterator(TreeNode root) {
        q = new LinkedList<>();
        dfs(root, q);
    }
    
    /** @return the next smallest number */
    public int next() {
        return q.poll().val;
    }
    
    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        return !q.isEmpty();
    }
    private void dfs(TreeNode root, Queue<TreeNode> q) {
        if(root == null) {return;}
        dfs(root.left, q);
        q.offer(root);
        dfs(root.right, q);
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: Medium/173. BST Iterator/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.stack = []

        self.helper(root)

    
    def helper(self, root: TreeNode) -> None:

        while root:
            self.stack.append(root)
            root = root.left
        

    def next(self) -> int:

        next_node = self.stack.pop()

        if next_node.right:
            self.helper(next_node.right)
        
        return next_node.val
        

    def hasNext(self) -> bool:
        return len(self.stack) > 0